src/ImporterService.php

```diff
@@ -0,0 +1,370 @@
+   *
+   * @return bool
+   *    A TRUE/FALSE value to be used by array_filter.
+   */
+  protected static function preSave($row) {
+    // Disallow redirects from <front>.
+    if ($row['source'] == '<front>') {
+      self::$messages['warning'][] = t('You cannot create a redirect from the front page. Bypassing "@source".', array('@source' => $row['source']));
+      return FALSE;
+    }
+
+    // Disallow redirects from anchor fragments.
+    if (strpos($row['source'], '#') !== FALSE) {
+      self::$messages['warning'][] = t('Redirects from anchor fragments (i.e., with "#) are not allowed. Bypassing "@source".', array('@source' => $row['source']));
+      return FALSE;
+    }
+
+    // Disallow redirects to nonexistent internal paths.
+    if (self::InternalPathMissing($row['redirect'])) {
+      self::$messages['warning'][] = t('The destination path "@redirect" does not exist on the site. Redirect from "@source" bypassed.', array('@redirect' => $row['redirect'], '@source' => $row['source']));
+      return FALSE;
+    }
+
+    // Disallow infinite redirects.
+    if (self::sourceIsDestination($row)) {
+      self::$messages['warning'][] = t('You are attempting to redirect "@redirect" to itself. Bypassed, as this will result in an infinite loop.', array('@redirect' => $row['redirect']));
+      return FALSE;
+    }
+
+    return TRUE;
+  }
+
+  /**
+   * Save an individual redirect entity, if no redirect already exists.
+   *
+   * @param str[] $redirect_array
+   *    Keyed array of redirects, in the format
+   *    [source, redirect, status_code, language].
+   * @param bool $override
+   *    A 1 indicates that existing redirects should be updated.
+   */
+  public static function save($redirect_array, $override) {
+    if ($redirects = self::redirectExists($redirect_array)) {
+      if ($override == 1) {
+        $redirect = reset($redirects);
+        $message_type = 'Updated';
+      }
+      else {
+        return;
+      }
+    }
+    else {
+      $message_type = 'Added';
+      $parsed_url = UrlHelper::parse(trim($redirect_array['source']));
+      $path = isset($parsed_url['path']) ? $parsed_url['path'] : NULL;
+      $query = isset($parsed_url['query']) ? $parsed_url['query'] : NULL;
+
+      /** @var \Drupal\redirect\Entity\Redirect $redirect */
+      $redirectEntityManager = \Drupal::service('entity.manager')->getStorage('redirect');
+      $redirect = $redirectEntityManager->create();
+      $redirect->setSource($path, $query);
+    }
+    // Currently, the Redirect module's setRedirect function assumes
+    // all paths are internal. Therefore, we set it two ways.
+    if (parse_url($redirect_array['redirect'], PHP_URL_SCHEME)) {
+      $redirect->redirect_redirect->set(0, array('uri' => $redirect_array['redirect']));
+    }
+    else {
+      $redirect->setRedirect($redirect_array['redirect']);
+    }
+    $redirect->setStatusCode($redirect_array['status_code']);
+    $redirect->setLanguage($redirect_array['language']);
```

Comment:
Should we maybe check if the langcode exists? See: https://github.com/BurdaMagazinOrg/module-path_redirect_import/pull/2/files#diff-20cbc968b7cf68080ccff37715893e43R111